test/e2e/fqdn_dns_cache_test.go

```diff
@@ -0,0 +1,318 @@
+// Copyright 2024 Antrea Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"bytes"
+	"fmt"
+	"strconv"
+	"strings"
+	"testing"
+	"text/template"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/ptr"
+
+	crdv1beta1 "antrea.io/antrea/pkg/apis/crd/v1beta1"
+	agentconfig "antrea.io/antrea/pkg/config/agent"
+	"antrea.io/antrea/test/e2e/utils"
+)
+
+func TestFQDNPolicyWithCachedDNS(t *testing.T) {
+	const (
+		testFQDN = "fqdn-test-pod.lfx.test"
+		dnsPort  = 53
+	)
+
+	skipIfAntreaPolicyDisabled(t)
+	skipIfNotIPv4Cluster(t)
+	skipIfIPv6Cluster(t)
+
+	data, err := setupTest(t)
+	if err != nil {
+		t.Fatalf("Error when setting up test: %v", err)
+	}
+	defer teardownTest(t, data)
+
+	// create two agnHost pods and get their IPv4 addresses. The IP of these pods will be mapped against the FQDN.
+	podCount := 2
+	agnHostPodIps := make([]*PodIPs, podCount)
+	for i := 0; i < podCount; i++ {
+		agnHostPodIps[i] = createHttpAgnHostPod(t, data)
+	}
+
+	// get IPv4 addresses of the agnHost pods created.
+	agnHostPodOneIP, _ := agnHostPodIps[0].AsStrings()
+	agnHostPodTwoIP, _ := agnHostPodIps[1].AsStrings()
+
+	// create customDNS service and get its ClusterIP.
+	customDnsService, err := data.CreateServiceWithAnnotations("custom-dns-service", data.testNamespace, dnsPort,
+		dnsPort, corev1.ProtocolUDP, map[string]string{"app": "custom-dns"}, false,
+		false, corev1.ServiceTypeClusterIP, ptr.To[corev1.IPFamily](corev1.IPv4Protocol), map[string]string{})
+	require.NoError(t, err, "Error creating customDNS service: %+v", err)
+	dnsServiceIP := customDnsService.Spec.ClusterIP
+
+	// create a ConfigMap for the custom DNS server, mapping IP of agnHost pod 1 to the FQDN.
+	dnsConfigData := createDnsConfig(t, map[string]string{"ipAddress": agnHostPodOneIP, "domainName": testFQDN})
+	configMap := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "custom-dns-config",
+			Namespace: data.testNamespace,
+		},
+		Data: dnsConfigData,
+	}
+	customDnsConfigMap, err := data.CreateConfigMap(configMap)
+	require.NoError(t, err, "failed to create custom dns ConfigMap: %v", err)
```

Comment:
```suggestion
	require.NoError(t, err, "failed to create custom dns ConfigMap")
```

`require.NoError` will already show the error, so adding it to the message just duplicates it in the output in case of error